src/dbxincluder/xinclude.py

```diff
@@ -0,0 +1,305 @@
+            # Override/Remove xml:id on all top-level elements
+            if value:
+                subtree.set(QN['xml:id'], value)
+            elif subtree.get(QN['xml:id']):
+                del subtree.attrib[QN['xml:id']]
+        elif qname.namespace == NS['local']:
+            # Set attribute on all top-level elements
+            subtree.set(qname.localname, value)
+        elif qname.namespace == NS['xml']:
+            # Ignore xml: namespace
+            continue
+        elif qname.namespace is not None:
+            # Set attribute on all top-level elements
+            subtree.set(name, value)
+
+
+def get_target(elem, base_url, file=None):
+    """Return tuple of the content of the target document as string and the URL that was used
+
+    :param elem: XInclude element
+    :param base_url: xml:base of the element
+    :raises DBXIException: href attribute is missing
+    :raises ResourceError: Couldn't fetch target
+    """
+
+    # Get href
+    href = elem.get("href")
+    if href is None:
+        raise DBXIException(elem, "Missing href attribute", file)
+
+    # Build full URL
+    urlparts = base_url.split("/")
+    if len(urlparts) > 1:
+        url = "/".join(urlparts[:-1]) + "/" + href
+    else:  # pragma: no cover
+        url = href
+
+    try:
+        if "://" in base_url:
+            target = urllib.request.urlopen(base_url)
+        else:  # Add file:// for URLs without scheme
+            target = urllib.request.urlopen("file://" + os.path.abspath(url))
+        content = target.read()
+        target.close()
+    except urllib.error.URLError:
+        raise ResourceError(elem, "Could not get target {0!r}".format(url), file)
+    except IOError as ioex:
+        raise ResourceError(elem, "Could not get target {0!r}: {1}".format(url, ioex), file)
+
+    return content, url
+
+
+def handle_xifallback(elem, file=None, xinclude_stack=None):
+    """Process the xi:include tag elem.
+    It will be replaced by the content of the xi:fallback subelement.
+
+    :param elem: The XInclude element to process
+    :param file: URL used to report errors
+    :param xinclude_stack: List (or None) of str with url and fragid to detect infinite recursion
+    :return: True if xi:fallback found"""
+
+    # There can be only xi:fallback in a xi:include, so just use the first child
+    if len(elem) == 0 or not isinstance(elem.tag, str) or QName(elem[0]) != QN['xi:fallback']:
+        return False
+
+    # Save the tailing text
+    append_to_tail(elem[0], elem.tail)
+
+    # process_tree before replacement to not lose xml:base on xi:include or xi:fallback
+    process_tree(elem[0], None, file, xinclude_stack)
+
+    # Two passes for fallback processing, flatten them after process_tree
+    elem.getparent().replace(elem, elem[0])
+    return True
+
+
+def validate_xinclude(elem, file):
+    """Raise DBXIException if the XInclude element elem is not valid."""
+
+    valid_attributes = ["href", "fragid", "parse", "set-xml-id"]
```

Comment:
Maybe use a tuple here to disallow further assignments? (Probably a matter of taste.)